### PR TITLE
[stable9] Filecache repair step

### PR DESCRIPTION
**Pull request by tomneedham**
Backport of https://github.com/owncloud/core/pull/28253 for stable9 (ignore the branch name, rebased for stable9)

### CODE REVIEW

**Comment by tomneedham** [comment]
Do we have console output access for repair steps? I just mapped it over to a OCP\ILogger instance for now

**Comment by tomneedham** [comment]
Needs a couple of fixes from https://github.com/owncloud/core/pull/28253 then this should work here too 🎉

**Comment by tomneedham** [comment]
Ok let's see how this goes - I think the tests might need some more tweaking with regards to output / repairstep differences

**Comment by tomneedham** [comment]
Ahh yes we dont have `createMock()`

**Comment by tomneedham** [comment]
detached subtree repair step fails on oracle...

**Comment by tomneedham** [comment]
Was missing the last fix from vincent https://github.com/owncloud/core/pull/28253/commits/4c157211685c450b46437b1188ed849676670617

**Comment by tomneedham** [comment]
Let's wait for CI here

**Comment by tomneedham** [comment]
Jenkins passed 💪

**Comment by PVince81** [comment]
@tomneedham is this up to date with the latest changes from the stable10 PR ?

before merging stable9 would be good to also have stable9.1 backport and merge that one first

**Comment by tomneedham** [comment]
Rebased and added path hash commit for performance

File: apps/files/command/scan.php [hunk 1]
```diff
@@ -198,9 +284,10 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 			if ($this->userManager->userExists($user)) {
 				# add an extra line when verbose is set to optical separate users
 				if ($verbose) {$output->writeln(""); }
-				$output->writeln("Starting scan for user $user_count out of $users_total ($user)");
+				$r = $shouldRepairStoragesIndividually ? ' (and repair)' : '';
```
Comment (PVince81) [inline apps/files/command/scan.php]: I don't remember these changes, did you add them to master/stable10 as well ?

File: lib/private/repair/repairmismatchfilecachepath.php [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+		$this->connection->beginTransaction();
+
+		$qb = $this->connection->getQueryBuilder();
+		$qb->update('filecache')
+			->set('parent', $qb->createNamedParameter($correctParentId))
+			->where($qb->expr()->eq('fileid', $qb->createNamedParameter($fileId)));
+		$qb->execute();
+
+		$text = "Fixed file cache entry with fileid $fileId, set wrong parent \"$wrongParentId\" to \"$correctParentId\"";
+		$out->advance(1, $text);
```
Comment (PVince81) [inline lib/private/repair/repairmismatchfilecachepath.php]: I didn't know we had the ability to display progress in stable9, good

File: lib/private/repair/repairmismatchfilecachepath.php [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+				// just fix all
+				$this->fixEntriesWithNonExistingParentIdEntry($out);
+			}
+			$out->finishProgress();
+			$out->info('');
+		}
+	}
+}
+
+class LogOutput {
```
Comment (PVince81) [inline lib/private/repair/repairmismatchfilecachepath.php]: Bäh... in my original backport I just removed progress logging.

**Review by PVince81 — COMMENTED** [review]
Changes look ok, but I feel slightly disturbed by the hard-coded LogOutput class. Maybe just remove all logging ?

The other alternative is to use the event emitter like we do in other repair steps. But that might not work when trying to log to the console when run from the scan command.

**Comment by PVince81** [comment]
Weird JS setup failures, same like https://github.com/owncloud/core/pull/29407#issuecomment-342280762

**Comment by PVince81** [comment]
Will require https://github.com/owncloud/core/pull/29477 for JS tests to pass and maybe a PR resubmit to clean the env.

**Comment by PVince81** [comment]
Resend as https://github.com/owncloud/core/pull/29489 to fix JS env issue

File: apps/files/command/scan.php (hunk 1 above)
Comment (tomneedham) [reply apps/files/command/scan.php]: master https://github.com/owncloud/core/blob/master/apps/files/lib/Command/Scan.php#L305
stable10 https://github.com/owncloud/core/blob/stable10/apps/files/lib/Command/Scan.php#L305

**Pull request closed without merging**